docs/halibut-protocol.md

```diff
@@ -0,0 +1,77 @@
+# Halibut Protocol data exchange
+
+This document outlines the data that is exchanged between the **Client** and **Service**. In this document:
+
+ - **Client** is what makes the remote procedure calls e.g. Octopus Server.
+ - **Service** is what executes the methods e.g. Tentacle. 
+
+
+When invoking a remote method in both polling and listening mode, the `Client` and `Service` go through three stages.
+
+ ### Stage 1 - Identification
+ 
+  A TCP connection is established and both server and client identify themselves, though a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.
+
+### Stage 2 - Message exchange
+
+Now that the connections are established, the Service will wait for requests from the Client. In both polling and listening the following steps are repeated in order:
+
+1. The Client sends a `Request` message containing the method to execute and data.
+2. The Service executes the method and sends the result in a `Response` message.
+3. The Service sends a `NEXT` control message, signaling it is ready for the next action.
+4. The Client sends a `PROCEED` control message, signaling to the Service to be ready for another `Request`.
+
+The steps are repeated for both Listening and Polling mode until either Client or Service begins the Ending a connection stage.
```

Comment:
Is it worth mentioning that during polling, the `NEXT`/`PROCEED` cycle repeats itself periodically when no requests are being sent?

Reply:
Done under the polling section